hephaistus/main-theme/distribution-system/chord/playground/nullist0/lib/src/main/kotlin/org/lia/dht/nullist0/chord/model/AbstractChordNode.kt

```diff
@@ -0,0 +1,65 @@
+package org.lia.dht.nullist0.chord.model
+
+import java.net.URI
+
+
+/**
+ * An abstract class to represent the node in a chord ring.
+ */
+abstract class AbstractChordNode<Id, Value> {
+    abstract val uri: URI
+    abstract val predecessorOrNull: AbstractChordNode<Id, Value>?
+    abstract val successorOrNull: AbstractChordNode<Id, Value>?
```

Comment:
How do you think change it to null-safety?

successor and predecessor updates when creating new node, join node to existed DHT or stabilizing DHT.
* Create DHT
  * Run in single node DHT
  * successor and predecessor can be this node
* Join
  * Add this node to another DHT
  * successor is target and predecessor is this
* Stabilize
  * Fix this DHT
  * It update predecessor of the successor

Reply:
I think the `predecessor` and `successor` should be nullable. If not, all chord node data should be contain all chord nodes, like doubly linked list. This will violate the benefit of DHT.

Reply:
Yeah, sounds nice approach ! It was the wrong direction what I made. BTW, why it's violate..?.? I couldn't get answer of it.. 😔

Reply:
Ah... I'm sorry for the ambigous words. I mean that the benefit of the Chord algorithm is holding less data about other nodes, so that large chord ring can be performed as well because there is a little nodes (O(logN)) to be held.

Reply:
I suggest it for robustness of Chord DHT. I imagine each node has strict reference than now, Chord DHT can be more robustness one. With nullable state, each node can be isolated from DHT. BTW, it increases complexity of context as you told.

O(logN) is achieved by the finger table and, IMHO, it's not related with my suggestion.

Reply:
Thanks for kind explanation. I will leave this now. 🙇